src/openrct2/ride/Vehicle.cpp

```diff
@@ -1830,9 +1831,26 @@ static void vehicle_update_measurements(rct_vehicle * vehicle)
     }
 
     rct_tile_element * tile_element = map_get_surface_element_at({x, y});
-    if (tile_element->base_height * 8 <= vehicle->z)
+    if (tile_element->base_height * 8 <= z) // If vehicle above ground.
     {
+        // Couldn't get this to work?
+        //tile_element = map_get_nth_element_at(x/32, y/32, z/8);
+
+        bool tile_element_at_ride_height = false;
+        // Iterate up the layers until you hit a piece of track from the same ride.
+        do
+        {
+            tile_element++;
+            if (tile_element_get_type(tile_element) == TILE_ELEMENT_TYPE_TRACK)
+            {
+                if (tile_element->properties.track.ride_index == vehicle->ride)
+                {
+                    tile_element_at_ride_height = true;
+                }
+            }
+        } while (!tile_element_at_ride_height);
```

Comment:
The logic here has a few flaws in it. When there are multiple track elements of the same ride on the same tile, it will always find the first element. The second flaw is that tile elements aren't necessarily ordered from bottom to top (though this is not achievable through normal gameplay), meaning the next loop will start iterating at a seemingly random element.

Instead of picking the tile element to start iterating from, you should start from the first element on the tile and at the beginning of the loop compare its base height with the vehicle's height (keep in mind that `base_height` uses small coordinates, so multiply this with 8 first). If it's lower, `continue`. Otherwise the rest of the loop will executed and `cover_found` will be set correctly.

Reply:
Thank you, I believe I've integrated the changes. Your comment was a great help.